src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.KeepAlive.cs

```diff
@@ -0,0 +1,316 @@
+            {
+                // If Exception is not null, it triggered the abort which also disposed the websocket
+                // We only save the Exception if it actually triggered the abort
+                if (exc is ObjectDisposedException ode && ode.ObjectName == typeof(ManagedWebSocket).FullName)
+                {
+                    throw new OperationCanceledException(nameof(WebSocketState.Aborted), _keepAlivePingState.Exception);
+                }
+
+                if (exc is WebSocketException we && we.WebSocketErrorCode == WebSocketError.InvalidState)
+                {
+                    throw new WebSocketException(WebSocketError.InvalidState, we.Message, _keepAlivePingState.Exception);
+                }
+            }
+        }
+
+        private sealed class KeepAlivePingState
+        {
+            internal const int PingPayloadSize = sizeof(long);
+            internal const long MinIntervalMs = 1;
+
+            internal long DelayMs;
+            internal long TimeoutMs;
+            internal long NextPingTimestamp;
+            internal long WillTimeoutTimestamp;
+
+            internal long HeartBeatIntervalMs;
+
+            internal bool AwaitingPong;
+            internal long PingPayload;
+            internal Exception? Exception;
+
+            public KeepAlivePingState(TimeSpan keepAliveInterval, TimeSpan keepAliveTimeout)
+            {
+                DelayMs = TimeSpanToMs(keepAliveInterval);
+                TimeoutMs = TimeSpanToMs(keepAliveTimeout);
+                NextPingTimestamp = Environment.TickCount64 + DelayMs;
+                WillTimeoutTimestamp = Timeout.Infinite;
+
+                HeartBeatIntervalMs = Math.Max(
+                    Math.Min(DelayMs, TimeoutMs) / 4,
+                    MinIntervalMs);
+
+                static long TimeSpanToMs(TimeSpan value) => Math.Max(
+                        (long) Math.Min(value.TotalMilliseconds, int.MaxValue),
+                        MinIntervalMs);
+            }
+
+            internal void OnDataReceived()
+            {
+                if (NetEventSource.Log.IsEnabled()) NetEventSource.Trace(this);
+
+                Interlocked.Exchange(ref NextPingTimestamp, Environment.TickCount64 + DelayMs);
+            }
+
+            internal void OnPongResponseReceived(Span<byte> pongPayload)
+            {
+                Debug.Assert(AwaitingPong);
+                Debug.Assert(pongPayload.Length == sizeof(long));
+
+                if (NetEventSource.Log.IsEnabled()) NetEventSource.Trace(this);
+
+                long pongPayloadValue = BinaryPrimitives.ReadInt64BigEndian(pongPayload);
+                if (pongPayloadValue == Interlocked.Read(ref PingPayload))
+                {
+                    if (NetEventSource.Log.IsEnabled()) NetEventSource.PongResponseReceived(this, pongPayloadValue);
+
+                    Interlocked.Exchange(ref WillTimeoutTimestamp, Timeout.Infinite);
+                    AwaitingPong = false;
```

Comment:
You have multiple pieces of state that are changing from multiple threads without synchronization.

I think there's race conditions here, e.g.:
- We sent a ping frame, `AwaitingPong = true`, `WillTimeoutTimestamp = 42`
- The timer thread checks `AwaitingPing` and calls into `KeepAlivePingThrowIfTimedOut`.
- A different thread observes a pong frame, sets `WillTimeoutTimestamp = -1`.
- The timer thread in `KeepAlivePingThrowIfTimedOut` blows up on debug asserts or checks `now > -1` and throws even though the peer was responding

I'd recommend moving all of this timing logic behind a lock to make it easier to prove correctness